op-program/host/host.go

```diff
@@ -241,50 +89,42 @@ func makeDefaultPrefetcher(ctx context.Context, logger log.Logger, kv kvstore.KV
 	l1BlobFetcher := sources.NewL1BeaconClient(l1Beacon, sources.L1BeaconClientConfig{FetchAllSidecars: false})
 
 	logger.Info("Initializing L2 clients")
-	l2Client, err := NewL2Source(ctx, logger, cfg)
+	l2Client, err := hostcommon.NewL2Source(ctx, logger, cfg)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create L2 source: %w", err)
 	}
 
-	return prefetcher.NewPrefetcher(logger, l1Cl, l1BlobFetcher, l2Client, kv), nil
+	executor := MakeProgramExecutor(logger, cfg)
+	return prefetcher.NewPrefetcher(logger, l1Cl, l1BlobFetcher, l2Client, kv, cfg.L2ChainConfig, executor), nil
 }
 
-func routeHints(logger log.Logger, hHostRW io.ReadWriter, hinter preimage.HintHandler) chan error {
-	chErr := make(chan error)
-	hintReader := preimage.NewHintReader(hHostRW)
-	go func() {
-		defer close(chErr)
-		for {
-			if err := hintReader.NextHint(hinter); err != nil {
-				if err == io.EOF || errors.Is(err, fs.ErrClosed) {
-					logger.Debug("closing pre-image hint handler")
-					return
-				}
-				logger.Error("pre-image hint router error", "err", err)
-				chErr <- err
-				return
-			}
-		}
-	}()
-	return chErr
+type programExecutor struct {
+	logger log.Logger
+	cfg    *config.Config
 }
 
-func launchOracleServer(logger log.Logger, pHostRW io.ReadWriteCloser, getter preimage.PreimageGetter) chan error {
-	chErr := make(chan error)
-	server := preimage.NewOracleServer(pHostRW)
-	go func() {
-		defer close(chErr)
-		for {
-			if err := server.NextPreimageRequest(getter); err != nil {
-				if err == io.EOF || errors.Is(err, fs.ErrClosed) {
-					logger.Debug("closing pre-image server")
-					return
-				}
-				logger.Error("pre-image server error", "error", err)
-				chErr <- err
-				return
-			}
-		}
-	}()
-	return chErr
+func (p *programExecutor) RunProgram(
+	ctx context.Context,
+	prefetcher hostcommon.Prefetcher,
+	blockNum uint64,
+	chainID uint64,
+) error {
+	newCfg := *p.cfg
+	newCfg.L2ChainID = chainID
+	newCfg.L2ClaimBlockNumber = blockNum
+	// TODO(interop): Configure the chainID and RPC endpoints
```

Comment:
We are configuring the chain ID here already. I'd have thought we'd just inherit the RPC endpoints from the original config because we can't guess what they would be (we will need to update the config to support multiple endpoints for different chains eventually but that should just flow through here automatically).

Reply:
I don't yet know how multi-chain configs will work in the host. And I assumed that we'll need to manually configure RPCs somehow.
But you make a good point. The chainID should suffice for a `MultichainL2Fetcher` or whatnot to route RPC requests to the appropriate endpoint. So `ProgramExecutor` shouldn't be concerned with how the endpoints are configured.